gsdk/codegen/src/storage.rs

```diff
@@ -0,0 +1,131 @@
+use quote::{quote, ToTokens};
+use syn::{
+    parse_quote, punctuated::Punctuated, Expr, ExprLit, FnArg, Ident, ItemFn, Lit, LitStr, Meta,
+    MetaNameValue, PatType,
+};
+
+const FULL_DOC_SUFFIX: &str = " at specified block";
+const FULL_SUFFIX: &str = "_at";
+
+/// Storage query builder for generating
+/// - basic private shared query
+/// - full query with block hash
+/// - short query with block hash None.
+pub struct StorageQueryBuilder(ItemFn);
+
+impl StorageQueryBuilder {
+    /// Build storage query from full blocks.
+    fn full(&self) -> ItemFn {
+        let mut short = self.0.clone();
+
+        // reset function docs.
+        //
+        // - remove `at specified block` suffix.
+        short.attrs.iter_mut().find_map(|attr| {
+            if let Meta::NameValue(MetaNameValue {
+                value:
+                    Expr::Lit(ExprLit {
+                        attrs: _,
+                        lit: Lit::Str(lit_str),
+                    }),
+                ..
+            }) = &mut attr.meta
+            {
+                *lit_str = LitStr::new(
+                    &lit_str.value().replace(FULL_DOC_SUFFIX, ""),
+                    lit_str.span(),
+                );
+                return Some(());
+            }
+
+            None
+        });
+
+        // reset function name.
+        //
+        // - `storage_query_at` -> `storage_query`
+        short.sig.ident = Ident::new(
+            short.sig.ident.to_string().trim_end_matches(FULL_SUFFIX),
+            short.sig.ident.span(),
+        );
+
+        // reset function inputs.
+        //
+        // - remove `block_hash: Option<H256>`
+        short.sig.inputs = Punctuated::from_iter(short.sig.inputs.into_iter().filter(|v| {
+            if let FnArg::Typed(PatType { pat, .. }) = v {
+                return !pat.to_token_stream().to_string().contains("block_hash");
+            }
+
+            true
+        }));
+
+        // reset function block.
+        //
+        // ```ignore
+        // {
+        //   self.storage_query(..args, None);
+        // }
+        // ```
+        {
+            let fn_at = &self.0.sig.ident;
+            let args = short
+                .sig
+                .inputs
+                .iter()
+                .filter_map(|v| {
+                    if let FnArg::Typed(PatType { pat, .. }) = v {
+                        Some(Ident::new(
+                            &pat.to_token_stream().to_string(),
+                            Span::call_site(),
+                        ))
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<Ident>>();
+
+            short.block.stmts = parse_quote! {
+                self.#fn_at(#(#args,)* None).await
+            };
+        }
+
+        short
+    }
+
+    /// Build all storage queries.
+    pub fn build(&self) -> TokenStream {
+        let (at, full) = (self.0.clone(), self.full());
```

Comment:
I'd request to write original signature using Option<H256> and placing into impl within macro expand